### ISSUE TITLE
Импорт моделей из метробилда 375 ?

### ISSUE THREAD

**Issue by moderamodera** [PavelBlend/blender-xray#788]
Привет! Меня тут в комментах на ютубе попросили написать авторам плагина blender-xray насчёт импорта моделей (а может даже и анимации) из единственного пока доступного билда метро 2033

Не уверен что это нужно, т.к. плагин про сталкер и X-Ray, а тот билд официально ни тем ни другим не является. С другой стороны форматы ну очень уж аналогичны, изобретать отдельный плагин для них смысла нет. Отличия лишь такие:
1. У моделей чанк с вершинами имеет другой номер (5) и другой формат вершин, в точности как в метро 2033, всё остальное не менялось
2. У анимации ключи перемещения сделали 16-битными всегда. Как в ЧН, только без флажка указывающего 16-битность.

В принципе думаю я могу попробовать сделать форк с такой функцией, но поддерживать его в актуальном состоянии мне точно будет лень потом. 
![изображение](https://github.com/user-attachments/assets/577c91f8-e409-4466-ad6d-566676558346)

**Comment by PavelBlend** [comment]
@moderamodera Привет.

Не знаю, подумаю. Я пока аддоном не занимаюсь.

Потом можно будет сделать, если не много править придётся. Там ведь один формат? Или как в сталкере, несколько: 1/2/3/4-веса на вершину и т. д.

И у меня нет файлов для примера.

И что насчёт экспорта? Там ведь придётся добавлять параметры у моделей/костей, которых нет в сталкере. Как-то не хочется добавлять дополнительные свойства в панели и в окошки аддона.

**Comment by moderamodera** [comment]
@PavelBlend Да, формат вертексов один, всегда 4 кости на вертекс. Модели с билда прикрепляю.
[meshes_375.zip](https://github.com/user-attachments/files/16556147/meshes_375.zip)

Насчёт экспорта, эти форматы никаких новых параметров не несут по сравнению с форматами сталкера, достаточно будет добавить новый пункт в поле выбора версии. Или отдельный тип файлов в меню Import/Export. Однако делать экспорт никто пока не просил, можно этот вопрос отложить то тех пор пока кто-нибудь не попросит.

**Comment by PavelBlend** [comment]
@moderamodera привет. У тебя есть дискорд?

**Comment by moderamodera** [comment]
> @moderamodera привет. У тебя есть дискорд?

да, мой дискорд `modera.`

**Comment by PavelBlend** [comment]
@moderamodera отправил тебе заявку в друзья.

**Comment by moderamodera** [comment]
> @moderamodera отправил тебе заявку в друзья.

что-то пока не появилась

**Comment by moderamodera** [comment]
> @moderamodera отправил тебе заявку в друзья.

попробуй другой дискорд: `saas2838`
мне на тот что-то заявка не приходит

**Comment by PavelBlend** [comment]
Я точку в имени не поставил.

**Comment by PavelBlend** [comment]
@moderamodera привет.

Я сейчас пытался сделать импорт этих моделей. У меня не получается импортировать вершины. Как они устроены? Как вообще устроен этот чанк 5 с вершинами?

**Comment by moderamodera** [comment]
> [@moderamodera](https://github.com/moderamodera) привет.
> 
> Я сейчас пытался сделать импорт этих моделей. У меня не получается импортировать вершины. Как они устроены? Как вообще устроен этот чанк 5 с вершинами?

Выражаясь языком C, чанк устроен так:

```
struct SkinVertex {
    signed short xyzw[4]; // делить на 2720.f
    unsigned char normal[4]; // нормали закодированы так же как в уровнях сталкера
    unsigned char tangent[4];
    unsigned char binormal[4];
    unsigned char bones[4]; // делить на 3, индекс в массив SkinVerticesChunk.bones
    unsigned char weights[4]; // делить на 255.f
    signed short uv[2]; // делить на 2048.f
};

struct SkinVerticesChunk {
    unsigned char bone_count;
    unsigned char bones[bone_count];
    unsigned int vertex_count;
    struct SkinVertex[vertex_count]
};
```
